### ISSUE TITLE
No valid convolution algorithms available in CuDNN

### ISSUE THREAD

**Issue by djimenezsanchez** [DIAGNijmegen/pathology-streaming-pipeline#6]
Dear Hans, 

First of all, I would like to congratulate you on your work, I'm impressed by the methodology and results!

I'm happily adapting your neural network to another pathology imaging modality with the purpose of classifying patients to clinical subject-level labels. Due to the chosen imaging modality, I cannot use the pretrained ResNet that is chosen as the default configuration in your GitHub. Alternatively, I have to train all neural network layers, or, in other words, I need to set train_streaming_layer=True and train_all_layers=True.

When these parameters are set to True, unfortunately, the running process stops and an exception is raised at cpp_functions.backward function (line 221 from scnn.py):

Exception trace:
no valid convolution algorithms available in CuDNN
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/torch_utils/checkpointed_trainer.py", line 174, in backward_batch
    if self.mixedprecision: self.grad_scaler.scale(loss).backward()
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/torch_utils/checkpointed_trainer.py", line 124, in train_checkpointed_batch
    self.backward_batch(loss, fmap)
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/torch_utils/checkpointed_trainer.py", line 114, in train_batch_if_needed
    self.train_checkpointed_batch()
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/torch_utils/checkpointed_trainer.py", line 77, in train_full_dataloader
    self.train_batch_if_needed()
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/torch_utils/trainer.py", line 171, in train_epoch
    self.train_full_dataloader(batch_callback)
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/train_plex.py", line 145, in train_epoch
    preds, gt = trainer.train_epoch(self._train_batch_callback)
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/train_plex.py", line 139, in train_and_eval_epochs
    self.train_epoch(e, self.trainer)
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/train_plex.py", line 89, in run_experiment
    else: self.train_and_eval_epochs()
  File "/gpu-data/djsanchez/pathology-streaming-pipeline/train_plex.py", line 533, in <module>
    exp.run_experiment()

I understand that something is happening with the cpp_functions, but I cannot tell what, and I don't know how to continue. For this reason, I would really appreciate your help solving this issue and helping me out to run your neural network to train all layers.

Thanks in advance,
Daniel Jiménez-Sánchez.

**Comment by hanspinckaers** [comment]
Hi Daniel,

Can you try setting the env variable `CUDA_LAUNCH_BLOCKING=1`?

Also, it seems to happen in the second stage of your neural network (the not streaming part). How does that part of the network look like?

It can also happen with streaming when your tile size is not large enough, but then the error should not happen at this stage but earlier. 

Cheers,
Hans

**Comment by djimenezsanchez** [comment]
Hello Hans,

I set the env variable CUDA_LAUNCH_BLOCKING=1 as suggested, but unfortunately, the network is not working and the 'no valid convolution algorithms available in CuDNN' error is still raised. I created a fresh start with a new Conda environment, but still, I'm getting the same error.

I am using the default network with the only exception of changing train_all_layers=True, and pretrained=False. I tried different patch sizes, one of them being as big as 1800, but got the same error. 

Do you have any further suggestions?

Thanks!
Daniel Jiménez-Sánchez.

**Comment by hanspinckaers** [comment]
Hi Daniel,

I’m not working today so can’t check the code myself. Can take a look
tomorrow. Which platform are you? Does normal pytorch code train without
problems?

Thanks,
Hans

**Comment by djimenezsanchez** [comment]
Hi,

Yes, I will give you all the information concerning my setup. It worked with several PyTorch projects, but I never used custom .cpp PyTorch functions. 

Working environment:
PyTorch version: 1.6.0
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A
OS: Ubuntu 16.04.6 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Libc version: glibc-2.17
Python version: 3.6.13 |Anaconda, Inc.| (default, Jun  4 2021, 14:25:59)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.4.0-150-generic-x86_64-with-debian-stretch-sid
Is CUDA available: True
CUDA runtime version: 10.1.168
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
Versions of relevant libraries:
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py36he8ac12f_0  
[conda] mkl_fft                   1.3.0            py36h54f3939_0  
[conda] mkl_random                1.1.1            py36h0573a6f_0  
[conda] numpy                     1.19.2           py36h54aff64_0  
[conda] numpy-base                1.19.2           py36hfa32c7d_0  
[conda] pytorch                   1.6.0           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.7.0                py36_cu101    pytorch

I hope this helps!

**Comment by hanspinckaers** [comment]
I only see a discrepancy in cudnn versions between pytorch built and
installed. However, not sure if this would be a problem. Does regular
Pytorch code work in the same environment?

Thanks,
Hans

**Comment by hanspinckaers** [comment]
Hi Daniel,

Sorry, I didn't have time anymore, I will go on vacation on Monday, so will not be able to help hands-on. `CUDA_LAUNCH_BLOCKING=1` sometimes makes the error message more clear. 

You could maybe try a more recent PyTorch version.

Cheers,
Hans

**Comment by djimenezsanchez** [comment]
Hi Hans, 

I would like to reply to your previous suggestions:

- Does regular Pytorch code work in the same environment?
+ Yes regular PyTorch code is working. Indeed, if the model is trained on the last layers it works perfectly.

- You could maybe try a more recent PyTorch version.
+ To do so I need to update all drivers. I would need some extra time to do so because I'm using our GPU server to carry out other experiments.

If you have any other suggestions, please write.

Thank you very much for your efforts!
Daniel Jiménez.
